// mgsmu-react@1.0.5/dist/UseStore.js
import { useEffect, useState, useCallback } from 'react';
let state = {};
const listeners = new Map();
// Get the full state
export const getState = () => state;
// Set or update a specific key in the state
export const setStateStore = (entry, key = "data") => {
    var _a;
    state = Object.assign(Object.assign({}, state), { [key]: Object.assign(Object.assign({}, entry), { updatedAt: Date.now() }) });
    (_a = listeners.get(key)) === null || _a === void 0 ? void 0 : _a.forEach((listener) => listener());
};
// Remove one or multiple keys from the state
export const removeStateStore = (...keys) => {
    const newState = Object.assign({}, state);
    keys.forEach((key) => {
        delete newState[key];
    });
    state = Object.assign(Object.assign({}, newState), { updatedAt: Date.now() });
    listeners.forEach((set) => set.forEach((listener) => listener()));
};
// Subscribe to changes on a specific key
export const subscribeKey = (key, listener) => {
    if (!listeners.has(key))
        listeners.set(key, new Set());
    listeners.get(key).add(listener);
    return () => listeners.get(key).delete(listener);
};
// React hook to use the state
export const useStateStore = (key) => {
    const [value, setValue] = useState(state[key]);
    useEffect(() => {
        const update = () => setValue(state[key]);
        const unsubscribe = subscribeKey(key, update);
        return () => {
            unsubscribe();
        };
    }, [key]);
    const setGlobalValue = (v) => {
        const newValue = typeof v === "function" ? v(state[key]) : v;
        setStateStore(newValue, key);
    };
    return [value, setGlobalValue, removeStateStore];
};
// --- New click-only setter hook (write-only) ---
export const useClickSetter = (key) => {
    return useCallback((value) => {
        setStateStore(value, key);
    }, [key]);
};
